[P081_CRON] Fix conversion of time_t

clumsy-stefan opened the pull request:
time_t is 64 bit and gets cripled when converting to 32bit float. converting to a 64 bit long long fixes it.

clumsy-stefan committed:
Fix conversion of time_t

time_t is 64 bit and gets cripled when converting to 32bit float. converting to a 64 bit long long fixes it.

clumsy-stefan commented:
I was playing around with corn and found that the time's completely cripled, next execution time was evaluated to something like 2156-06-20 or so... I found that changing the conversion fixes it.
However there is stil an issue when displaying the converted date/time, it shows december as "00":
![image](https://user-images.githubusercontent.com/32835100/103177159-3f49ab00-4878-11eb-8e85-bb1c94735617.png)

Coulnd not yet find why...

uzi18 commented:
@clumsy-stefan @TD-er you can find easy fix here: #3441

clumsy-stefan commented:
Ok, this solves only part of the problem... the last and next execution times are stored in userVars, which are float (and therefore 32bit), so the plugin needs to be adapted to store the time values in 4 userVars (2 float's for each of the saved times). I'm having a look at it , but did not succeed yet...

uzi18 commented on src/_P081_Cron.ino:
```diff
@@ -77,7 +77,7 @@ private:
 {
   time_t time;
-  float  value;
+  long long  value;
```
@clumsy-stefan this should go into:
```
union timeToFloat
{
  time_t time;
  float  value;
  long long  value32b;
};
```

TD-er replied:
This is rather tricky, as those don't have the same size.
`float` is 32 bit, longlong (why not uint64_t?), which makes it rather compiler dependent if this will work as expected.

clumsy-stefan replied:
yes, that's why I changed it into a float value[2], like this it seems to work smoothly.. I'll update the PR later.. still looking why the date is printed wrongly...

clumsy-stefan commented:
I think that's not enough yet, as also
```
#define LASTEXECUTION UserVar[event->BaseVarIndex]
#define NEXTEXECUTION UserVar[event->BaseVarIndex+1]
```
will not work... everywhere in the code..

therefore I use now
```
union timeToFloat
{
  time_t time;
  float value[2];
};
```

and handle low 2 bytes and high 2 byte fot the time values sepparately throughout the code, which seems to work now. except that the timeconversion still gives 2021-00 instead of 2020-12... but I'm onto that...

EDIT: and also need to increase the numberof userVars to 4, to store the full 64bit value for both times (total 128bit)..

clumsy-stefan commented:
I have now a fully and correct working version of the plugin, just the time display is somehow wrong... also when I print the current time 
`ESPEasy_time::getDateTimeString(*gmtime(&current_time))` shows `2021-00-27 21:35:29`. So it seems there is a more fundamental bug here...

uzi18 commented:
time_t is 4 bytes (32bits) on both esp8266 and esp32

clumsy-stefan commented:
hmm... interesting, when i print a `sizeof()` it tells me 8 bytes...

TD-er commented:
Using a union for conversion is asking for trouble.
I think we should rather have a generic function to read the data in the `UserVar` also as int value, so you can also store specific values instead of keep running into the same issues over and over again.
However this is a bit tricky as a lot of code just reads the UserVar array as floats without considering it may hold anything else.

uzi18 commented:
`  static_assert(sizeof(time_t)==4," not 4bytes");` does not throw error, 
but `static_assert(sizeof(time_t)==8," not 8bytes");`does

@TD-er as long as this union is used only inside one plugin it shouldn't be so tricky

clumsy-stefan commented:
here's the opposite:
![image](https://user-images.githubusercontent.com/32835100/103180118-18e53900-4893-11eb-93d7-cbd87b6ea1b0.png)

clumsy-stefan committed:
[P081_Cron] change to 64bit time_t

cahnged the plugin to use 64bit time_t and store values in 4 instead of 2 user variables.

uzi18 commented:
@TD-er @clumsy-stefan so this is first difference between linux and win32 compiler 
```
src/src/Helpers/ESPEasy_checks.cpp: In function 'void run_compiletime_checks()':
src/src/Helpers/ESPEasy_checks.cpp:100:3: error: static assertion failed: not 8bytes
   static_assert(sizeof(time_t)==8,"not 8bytes");

```

clumsy-stefan commented:
Could be, I'm using ArduinoIDE on a Mac 64bit with avr-gcc....

TD-er commented:
> @TD-er @clumsy-stefan so this is first difference between linux and win32 compiler
> 
> ```
> src/src/Helpers/ESPEasy_checks.cpp: In function 'void run_compiletime_checks()':
> src/src/Helpers/ESPEasy_checks.cpp:100:3: error: static assertion failed: not 8bytes
>    static_assert(sizeof(time_t)==8,"not 8bytes");
> ```

I have seen multiple instances where there is a difference between the compilers.

`time_t` is defined here as such in framework-arduinoespressif8266/tools/sdk/libc/..../types.h:
```c++
#define	_TIME_T_	long			/* time() */
....
#ifndef __time_t_defined
typedef _TIME_T_ time_t;
#define __time_t_defined
#endif
```
So if it is not 4 bytes then `__time_t_defined` was set somewhere else.

clumsy-stefan commented:
using extensa compiler in `include/sys/_types.h` the following is defined:
```
#if defined(_USE_LONG_TIME_T) || __LONG_MAX__ > 0x7fffffffL
#define _TIME_T_ long
#else
#define _TIME_T_ __int_least64_t
#endif
```
This is set (or actuall not set) in `include/newlib.h`:
```
/* Define to use type long for time_t.  */
/* #undef _WANT_USE_LONG_TIME_T */
```

newlib seems to be the default that is included in this version of the compiler...

So I guess by setting `_USE_LONG_TIME_T` in the above header file it would take again the long variant of time_t...

I wonder what else is "different" in this newlib compared to the standard lib...

EDIT: still this obviously does not solve the issue that the month is used wrongly in the time structures all over the place...
